site/source/tutorials/005_form_widgets/index.md

```diff
@@ -1,26 +1,159 @@
+			v('fieldset', { classes: this.classes(css.nameField) }, [
+				v('legend', { classes: this.classes(css.nameLabel) }, [ 'Name' ]),
+				w(TextInput, {
+					key: 'input1',
+					label: {
+						content: 'First Name',
+						hidden: true
+					},
+					placeholder: 'Given name',
+					type: 'text' as 'text'
+				}),
+				w(TextInput, {
+					key: 'input2',
+					label: {
+						content: 'Last Name',
+						hidden: true
+					},
+					placeholder: 'Family name',
+					type: 'text' as 'text'
+				})
+			]),
+			w(TextInput, {
+				key: 'input3',
+				label: 'Email address',
+				type: 'email'
+			}),
+			w(Button, {
+				content: 'Save',
+				type: 'submit'
+			})
+		]);
+	}
+```
+
+At this point, the user interface for the form is available, but it doesn't do anything since we haven't specified any event handlers. In the [last](../004_user_interactions/) tutorial, we learned how to add event handlers to custom widgets by assigning a method to an event. When using pre-built widgets, however, we pass the handlers as properties. For example, we are going to need a way to handle each text fields `change` event. To do that, we provide the desired handler function as the `onChange` property that is passed to the widget. Update the `render` method once again:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/WorkerForm.ts' lines:33-110 %}
+
+This form of the `render` method now does everything that we need it to: it creates the user interface and registers the event handlers that will update the application as the user enters information.
+
+The method starts by decomposing the properties down in to local constants. We still need to define those properties, so update the `WorkerFormProperties` interface to include them:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/WorkerForm.ts' lines:9-19 %}
+
+There are three types of properties that we are using in this form. The `firstName`, `lastName` and `email` properties are going to set the values that are displayed in the form fields. The `firstNameInvalid`, `lastNameInvalid` and `emailInvalid` properties will be used to indicate if a field contains invalid data (e.g. an invalid email address in the `email` field). Finally, the `onChange`, `onBlur` and `onSubmit` properties expose the events that the `WorkerForm` widget can emit. To see how these different property types are used, let's examine the properties that are being passed into first `TextInput` widget:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/WorkerForm.ts' lines:51-69 %}
+
+The first thing that we see is a `key` property. As mentioned before, a key is necessary whenever the same type of widget or virtual DOM element will be rendered by a widget. After that comes the `invalid` property. Setting this property to `true` will update the widget's appearance to indicate that the widget contains invalid data. Next comes the the `label`, `placeholder`, and `type` fields which map to the expected properties.
```

Comment:
"Next comes the the `label`, `placeholder`, and `type` fields which map to the expected properties." -> "The `label`, `placeholder`, and `type` fields map to their expected properties."